qtgui/QtPixelator.cpp

```diff
@@ -0,0 +1,243 @@
+{
+    // gauge is 10cm, so there will be a rectangle totaling a size of in_width*in_stitchesPerGauge/10 x in_height*in_rowsPerGauge/10 stixels,
+    // where each stixel will have a width and height to be computed from the IRL size of in_stitchesPerGauge/10cm x in_rowsPerGauge/10cm:
+    // the first goal, therefore, must be to figure out how to int-ify "stixel" width and height from that aspect ratio, 
+    // i.e. making them fractions of the least common multiple with the same aspect ratio as the actual stitch
+    // the total size in stixels is therefore the number of stixels (as computed from the first equation) times the correspoding stixel size.
+    unsigned stixelGcd{ greatest_common_divisor(in_rowsPerGauge, in_stitchesPerGauge) };
+    stitchWidth = in_stitchesPerGauge / stixelGcd;
+    stitchHeight = in_rowsPerGauge / stixelGcd;
+    stitchCount = (unsigned)(ceil((in_width / 10.) * in_stitchesPerGauge));
+    rowCount = (unsigned)(ceil((in_height / 10.) * in_rowsPerGauge));
+} 
+
+QImage QtPixelator::pixelate()
+{
+    QImage colorMap = imageBuffer.scaled(QSize(stitchCount, rowCount));
+    for (int y = 0; y < colorMap.height(); y++) {
+        QRgb* line = (QRgb*)colorMap.scanLine(y);
+        for (int x = 0; x < colorMap.width(); x++) {
+            // line[x] has an individual pixel
+            auto nearestColor = minDiff(QColor(line[x]), colors);
+            if (nearestColor.isValid()) line[x] = nearestColor.rgb();
+            else line[x] = Qt::black;
+        }
+        logging::LogStream::instance().getLogStream(logging::Level::DEBUG) << "x";
+    }
+    logging::LogStream::instance().getLogStream(logging::Level::DEBUG) << std::endl;
+    logging::LogStream::instance().getLogStream(logging::Level::DEBUG) << "Get pixelation template of size " << colorMap.width() << "x" << colorMap.height() << std::endl;
+    return colorMap;
+}
+
+bool QtPixelator::scalePixels(const QImage& colorMap)
+{
+    if (! ((colorMap.width() == stitchCount) && (colorMap.height() == rowCount))) return false;
+    imageBuffer = imageBuffer.scaled(QSize(stitchCount * stitchWidth, rowCount * stitchHeight));
+    QPainter qPainter(&imageBuffer);
+    for (int y = 0; y < colorMap.height(); y++) {
+        QRgb* line = (QRgb*)colorMap.scanLine(y);
+        for (int x = 0; x < colorMap.width(); x++) {
+            qPainter.setPen(QColor(line[x]));
+            qPainter.setBrush(QColor(line[x]));
+            qPainter.drawRect(x * stitchWidth, y * stitchHeight, stitchWidth, stitchHeight);
+        }
+        logging::LogStream::instance().getLogStream(logging::Level::DEBUG) << "x";
+    }
+    logging::LogStream::instance().getLogStream(logging::Level::DEBUG) << std::endl;
+    qPainter.end();
+    logging::LogStream::instance().getLogStream(logging::Level::DEBUG) << "Pixelation complete"<< std::endl;
+    return true;
+}
+
+errors::Code QtPixelator::checkSettings()
+{
+    if (imageBuffer.isNull())
+    {
+        return errors::WRONG_INPUT_FILE;

```

Comment:
either this style or the one inside `run`. But both at a time is really hard to read.